errors: minor (SystemError) refactoring

BridgeAR opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14531/

jasnell commented on lib/internal/errors.js:
```diff
@@ -98,7 +98,6 @@ function sysErrorMessage(prefix, ctx) {
 // and may have .path and .dest.
 class SystemError extends Error {
   constructor(key, context = {}) {
-    context = context || {};
```
FWIW, This is not really a duplicated default. This is there to catch the possibility of something like `new SystemError('foo', null)`, which is not handled by the default argument.

joyeecheung replied:
The current way `SystemError`s are instantiated in the code base warrants that `context` would at least be a `{}`, theoretically we don't even need to handle the default argument if we are being careful with new `SystemError`s.

Trott commented:
-1 to fast-tracking.

The edge case @jasnell describes likely warrants a test case.

BridgeAR commented:
I pushed another commit to remove the default values completely. They were not used at the moment. I also removed spread arguments that were unnecessary and was able to remove the default arguments in `message`. I renamed that function to `getMessage` to make clear that it is actually a function and a getter.

It is actually better not to have any default values for `SystemError` because that way we will detect faulty implementations early.

I also inlined a function that was only there to generate the message for `SystemError`s and I found some dead code in a test. I kept it there because it seems to mirror some internal functionality.

BridgeAR commented:
@joyeecheung @jasnell @addaleax PTAL

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14540/

BridgeAR commented:
Rebased due to conflicts.

New CI https://ci.nodejs.org/job/node-test-pull-request/14583/

BridgeAR commented:
Landed in 109cfa1511cb504866c44314d92552285a236b1c

nelsonlarocca commented:
From: Ruben Bridgewater <notifications@github.com>
Sent: Monday, April 30, 2018 7:56 AM
To: nodejs/node <node@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nodejs/node] errors: minor (SystemError) refactoring (#20337)


Closed #20337<https://github.com/nodejs/node/pull/20337>.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/nodejs/node/pull/20337#event-1601685648>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AcvRBOPkqCP_vmbRxAxvj8EZn5JKNRWFks5ttyYQgaJpZM4TmPtq>.
